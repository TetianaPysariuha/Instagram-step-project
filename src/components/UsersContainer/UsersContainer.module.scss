.userContainer{
    width: 100%;
    max-width: 300px;
    min-width: 100px;
    border: 3px solid rgba($color: gray, $alpha: 0.3);
    border-radius: 10px;
    margin: 20px 0;
    max-height: 50vh;
    overflow: scroll;
}
.user{
    display: grid;
    grid-template-columns: 1fr 1fr;
    & button{
        background-color: white;
        border: none;
        color: rgba($color: blue, $alpha: 0.8);
    }
}
