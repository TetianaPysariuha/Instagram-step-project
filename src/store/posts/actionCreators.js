import {
  GET_POSTS, GET_POST_BY_ID, GET_POSTS_BY_USER_ID, ADD_POST, UPDATE_POST, DELETE_POST, SHOW_MORE_CHANGE, GET_NEW_PAGE_POSTS, CLEAR_POSTS, GET_FAVORITE_POSTS, CHANGE_ISOPENPOST,
} from './actiions';

const SERVER_URL = 'https://nameless-lake-66137.herokuapp.com';

export const getPosts = () => async (dispatch) => {
  const posts = await fetch(`${SERVER_URL}/posts`).then((res) => res.json()).then((data) => data.data);
  dispatch({ type: GET_POSTS, payload: posts });
};

export const clearPosts = () => ({ type: CLEAR_POSTS });

export const loadNewPagePosts = (payload) => async (dispatch) => {
  if (payload.start !== null && payload.end) {
    const response = await fetch(`${SERVER_URL}/posts/quantity/${payload.start}&${payload.end}`).then((res) => res.json()).then((data) => data);
    dispatch({ type: GET_NEW_PAGE_POSTS, payload: { data: response.data, countAll: response.totalCount } });
  } else {
    console.error(`For loafing posts should be start, end and userId parametres. Received: start=${payload.start},  end=${payload.end}`);
  }
};

export const getPostById = (payload) => async (dispatch) => {
  const post = await fetch(`${SERVER_URL}/posts/${payload}`).then((res) => res.json()).then((data) => data.data);
  dispatch({ type: GET_POST_BY_ID, payload: post });
};

export const getPostsByUserId = (payload) => async (dispatch) => {
  const post = await fetch(`${SERVER_URL}/posts/user/${payload}`).then((res) => res.json()).then((data) => data.data);
  dispatch({ type: GET_POSTS_BY_USER_ID, payload: post });
};

export const addNewPost = (payload) => async (dispatch) => {
  const post = await fetch(`${SERVER_URL}/posts`, {
    method: 'POST',
    mode: 'cors',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload),
  }).then((res) => res.json()).then((data) => data.data);
  dispatch({ type: ADD_POST, payload: post });
};

export const updatePost = (payload) => async (dispatch) => {
  const post = await fetch(`${SERVER_URL}/posts/${payload.id}`, {
    method: 'PUT',
    mode: 'cors',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload.data),
  }).then((res) => res.json()).then((data) => data);
  dispatch({ type: UPDATE_POST, payload: post });
};

export const deletePost = (payload) => async (dispatch) => {
  const post = await fetch(`${SERVER_URL}/posts/${payload}`, {
    method: 'DELETE',
    mode: 'cors',
    headers: { 'Content-Type': 'aplication/json' },
  }).then((res) => res.json()).then((data) => data.data);
  dispatch({ type: DELETE_POST, payload: post });
};

export const showMoreChange = (payload) => ({ type: SHOW_MORE_CHANGE, payload });

export const getFavoritePostsByUserId = (payload) => async (dispatch) => {
  const posts = await fetch(`${SERVER_URL}/posts/favorite/${payload}`).then((res) => res.json()).then((data) => data.data);
  dispatch({ type: GET_FAVORITE_POSTS, payload: posts });
};

export const changeIsOpenPost = (payload) => ({ type: CHANGE_ISOPENPOST, payload });
