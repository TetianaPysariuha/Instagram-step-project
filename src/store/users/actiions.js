export const LOGGING_USER = 'LOGGING_USER';
export const GET_USERS = 'GET_USERS';
export const GET_USER_BY_ID = 'GET_USER_BY_ID';
export const ADD_USER = 'ADD_USER';
export const UPDATE_USER = 'UPDATE_USER';
export const DELETE_USER = 'DELETE_USER';
export const GET_SUBSCRIBERS_BY_USER_ID = 'GET_SUBSCRIBERS_BY_USER_ID';
