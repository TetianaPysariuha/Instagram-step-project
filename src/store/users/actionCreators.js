import {
  LOGGING_USER, GET_USERS, GET_USER_BY_ID, ADD_USER, UPDATE_USER, DELETE_USER, GET_SUBSCRIBERS_BY_USER_ID,
} from './actiions';

const SERVER_URL = 'https://nameless-lake-66137.herokuapp.com';

export const loggedUserReceiveData = (payload) => async (dispatch) => {
  const user = await fetch(`${SERVER_URL}/users/${payload}`).then((res) => res.json()).then((data) => data.data);
  dispatch({ type: LOGGING_USER, payload: user });
};

export const getUsers = () => async (dispatch) => {
  const users = await fetch(`${SERVER_URL}/users`).then((res) => res.json()).then((data) => data.data);
  dispatch({ type: GET_USERS, payload: users });
};

export const getUserById = (payload) => async (dispatch) => {
  const user = await fetch(`${SERVER_URL}/users/${payload}`).then((res) => res.json()).then((data) => data.data);
  dispatch({ type: GET_USER_BY_ID, payload: user });
};

export const addNewUser = (payload) => async (dispatch) => {
  const user = await fetch(`${SERVER_URL}/users`, {
    method: 'POST',
    mode: 'cors',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload),
  }).then((res) => res.json()).then((data) => data.data);
  dispatch({ type: ADD_USER, payload: user });
};

export const updateUser = (payload) => async (dispatch) => {
  const users = await fetch(`${SERVER_URL}/users/${payload.id}`, {
    method: 'PUT',
    mode: 'cors',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload.data),
  }).then((res) => res.json()).then((data) => data.data);
  dispatch({ type: UPDATE_USER, payload: users });
};

export const deleteUser = (payload) => async (dispatch) => {
  const users = await fetch(`${SERVER_URL}/users/${payload}`, {
    method: 'DELETE',
    mode: 'cors',
    headers: { 'Content-Type': 'aplication/json' },
  }).then((res) => res.json()).then((data) => data.data);
  dispatch({ type: DELETE_USER, payload: users });
};

export const getSubscribersByUserId = (payload) => async (dispatch) => {
  const subscribers = await fetch(`${SERVER_URL}/users/${payload}`).then((res) => res.json()).then((data) => data.data);
  dispatch({ type: GET_SUBSCRIBERS_BY_USER_ID, payload: subscribers });
};
