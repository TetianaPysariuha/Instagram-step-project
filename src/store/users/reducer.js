import {
  LOGGING_USER, GET_USERS, GET_USER_BY_ID, ADD_USER, UPDATE_USER, DELETE_USER, GET_SUBSCRIBERS_BY_USER_ID,
} from './actiions';

const initialState = {
  users: [],
  currentUser: {},
  loggedUser: {},
  subscribers: [],
};

// the rule is switched off because state shoul be first in props
// eslint-disable-next-line default-param-last
const usersReducer = (state = initialState, action) => {
  switch (action.type) {
    case LOGGING_USER: {
      // eslint-disable-next-line no-unused-vars
      const { posts, __v, ...other } = action.payload;
      return { ...state, loggedUser: other };
    }
    case GET_USERS: {
      return { ...state, users: action.payload };
    }
    case GET_USER_BY_ID: {
      return { ...state, currentUser: action.payload };
    }
    case ADD_USER: {
      const newUsers = [...state.users];
      newUsers.push(action.payload);
      return { ...state, users: newUsers };
    }
    case UPDATE_USER: {
      return { ...state, users: action.payload };
    }
    case DELETE_USER: {
      return { ...state, users: action.payload };
    }
    case GET_SUBSCRIBERS_BY_USER_ID: {
      return { ...state, subscribers: action.payload };
    }
    default: {
      return state;
    }
  }
};

export default usersReducer;
