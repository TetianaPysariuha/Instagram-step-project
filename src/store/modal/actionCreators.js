import { OPEN_MODAL, CLOSE_MODAL } from './actiions';

export const openModalAC = (payload) => ({ type: OPEN_MODAL, payload });
export const closeModalAC = () => ({ type: CLOSE_MODAL });
