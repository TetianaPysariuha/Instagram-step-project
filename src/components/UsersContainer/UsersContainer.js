/* eslint-disable no-underscore-dangle */
import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import styles from './UsersContainer.module.scss';
import UserName from '../UserName/UserName';
import { updateUser } from '../../store/users/actionCreators';

function UsersContainer({ isButton, users, loggedUser }) {
  const dispatch = useDispatch();

  const handleClickSubscribe = (userId) => {
    let newFollowBy;
    const followBy = [...loggedUser.followBy];
    if (followBy.includes(userId)) {
      newFollowBy = followBy.filter((el) => el !== userId);
    } else {
      newFollowBy = [...followBy];
      newFollowBy.push(userId);
    }
    dispatch(updateUser({ id: loggedUser._id, data: { followBy: newFollowBy } }));
  };

  const getUserName = (user) => (
    <div className={styles.user} key={user._id}>
      <UserName image={user.avatar} nickname={user.nik} />
      {isButton && <button data-testid="subscribeBtn" type="button" onClick={() => handleClickSubscribe(user._id)}>{loggedUser.followBy.includes(user._id) ? 'Відписатись' : 'Підписатися '}</button>}
    </div>
  );

  return (
    <div className={styles.userContainer}>
      <ul>
        {users.length > 0 && loggedUser && users.map((user) => getUserName(user))}
      </ul>
    </div>
  );
}

UsersContainer.prototype = {
  isButton: PropTypes.bool,
  users: PropTypes.arrayOf(PropTypes.shape({
    _id: PropTypes.string,
    avatar: PropTypes.string,
    nik: PropTypes.string,
    name: PropTypes.string,
    followBy: PropTypes.arrayOf(PropTypes.string),
    __v: PropTypes.number,
    posts: PropTypes.arrayOf(PropTypes.shape({
      _id: PropTypes.string,
      img: PropTypes.string,
      userid: PropTypes.string,
      description: PropTypes.string,
      likes: PropTypes.arrayOf(PropTypes.string),
      favorite: PropTypes.arrayOf(PropTypes.string),
      comments: PropTypes.arrayOf(PropTypes.shape({
        _id: PropTypes.string,
        userid: PropTypes.string,
        text: PropTypes.string,
      })),
      seeCount: PropTypes.number,
      __v: PropTypes.number,
    })),
  })),
  loggedUser: PropTypes.shape({
    _id: PropTypes.string,
    avatar: PropTypes.string,
    nik: PropTypes.string,
    name: PropTypes.string,
    followBy: PropTypes.arrayOf(PropTypes.string),
  }),
};

export default memo(UsersContainer);
