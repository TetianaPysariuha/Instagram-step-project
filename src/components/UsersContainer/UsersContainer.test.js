import { render, screen, fireEvent } from '@testing-library/react';
import { useDispatch } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import UsersContainer from './UsersContainer';
import * as actions from '../../store/users/actionCreators';
import '@testing-library/jest-dom';

jest.mock('react-redux');
const user = {
  _id: '635429e5461d59d5a106db66',
  avatar: 'https://res.cloudinary.com/dx84fdyhd/image/upload/v1667755687/instagram/avatar/1_wfz4lc.jpg',
  nik: 'Milda',
  name: 'Patrisia Mumamba',
  followBy: [
    '6355802bab5157181c8ed0d2',
    '6355802bab5157181c8ed0d4',
    '6355802bab5157181c8ed0d5',
    '6355802bab5157181c8ed0d3',
    '635ad2dece62ea7848f38ea9',
    '6355802bab5157181c8ed0da',
  ],
};

const users = [user];

describe('Post snapshot testing', () => {
  test('Should UserContainer to render where isButton with value true', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <UsersContainer
          isButton
          users={users}
          loggedUser={user}
        />
      </BrowserRouter>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('Should UserContainer to render where isButton with value false', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <UsersContainer
          isButton={false}
          users={users}
          loggedUser={user}
        />
      </BrowserRouter>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Subscribe buttons work', () => {
  test('Click on button Subscribe', () => {
    const dispatch = jest.fn();
    useDispatch.mockReturnValue(dispatch);

    const mockedUpdateUser = jest.spyOn(actions, 'updateUser');
    const mockedGetUserById = jest.spyOn(actions, 'getUserById');

    render(
      <BrowserRouter>
        <UsersContainer
          isButton
          users={users}
          loggedUser={user}
        />
      </BrowserRouter>,
    );
    fireEvent.click(screen.getByTestId('subscribeBtn'));
    expect(dispatch).toHaveBeenCalledTimes(2);
    expect(mockedUpdateUser).toHaveBeenCalledTimes(1);
    expect(mockedGetUserById).toHaveBeenCalledTimes(1);
  });
});
