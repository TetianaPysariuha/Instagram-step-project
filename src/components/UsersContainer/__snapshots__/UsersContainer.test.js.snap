// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post snapshot testing Should UserContainer to render where isButton with value false 1`] = `
<DocumentFragment>
  <div
    class="userContainer"
  >
    <ul>
      <div
        class="user"
      >
        <div
          class="userNameContainer"
        >
          <a
            class="avatarLink"
            href="/Milda/"
          >
            <img
              alt="avatar"
              class="avatar"
              src="https://res.cloudinary.com/dx84fdyhd/image/upload/v1667755687/instagram/avatar/1_wfz4lc.jpg"
            />
          </a>
          <div
            class="nicknameWrraper"
          >
            <div>
              <a
                class="nickname"
                href="/Milda/"
              >
                <span>
                  Milda
                </span>
              </a>
              <p
                class="comment"
              />
            </div>
          </div>
        </div>
      </div>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Post snapshot testing Should UserContainer to render where isButton with value true 1`] = `
<DocumentFragment>
  <div
    class="userContainer"
  >
    <ul>
      <div
        class="user"
      >
        <div
          class="userNameContainer"
        >
          <a
            class="avatarLink"
            href="/Milda/"
          >
            <img
              alt="avatar"
              class="avatar"
              src="https://res.cloudinary.com/dx84fdyhd/image/upload/v1667755687/instagram/avatar/1_wfz4lc.jpg"
            />
          </a>
          <div
            class="nicknameWrraper"
          >
            <div>
              <a
                class="nickname"
                href="/Milda/"
              >
                <span>
                  Milda
                </span>
              </a>
              <p
                class="comment"
              />
            </div>
          </div>
        </div>
        <button
          data-testid="subscribeBtn"
          type="button"
        >
          Підписатися 
        </button>
      </div>
    </ul>
  </div>
</DocumentFragment>
`;
